### DISCUSSION TITLE
[JS] passed by value 와 passed by reference

### DISCUSSION THREAD

**Discussion by WooGie911** [wanted-9th-3team/frontend-study#91] [프론트 기술 질문 공간]
JS의 passed by value 와 passed by reference 에 대해 아는 만큼 설명해주세요.

**Comment by michoball** [comment]
함수의 파라미터로 들어온 값이 원시값인지 참조값인지에 대한 개념같습니다. 

함수로 원시값이 들어온다면 해당 값을 선언한 쪽에서는 값이 변경되지 않고 함수를 통과한 값만 변합니다. 

하지만 참조값이 들어오면 해당 값은 함수안에서도 참조값을 바라보고 함수가 실행되기 때문에 선언한 쪽과 함수를 통과한 후의 값이 동일하게 변합니다.

**Comment by WooGie911** [comment]
값을 저장할때 전달하는 방식입니다.

원시 타입(primitive type)의 값은 값에 의한 전달 (passed by value) 방식으로 전달되며,
객체 타입의 값은 참조에 의한 전달 (passed by reference) 방식으로 전달됩니다.

또한 원시 타입의 값은 변경할 수 없는 값이므로 원시 타입의 값을 변경하려면 새로운 값을 재할당해야 합니다.
반면 객체 타입의 값은 변경할 수 있는 값이므로 함수의 매개변수로 객체 타입의 값을 전달하면 함수 내부에서 매개변수의 값을 변경하면 원본 값도 변경됩니다.

**Comment by javascriptbiggosoo** [comment]
primitive가 argument로서 넘어가면 기본적으로 copy가 넘어간 것이지만 object가 넘어가는건 copy가 아니라 같은 memory heap을 참조하고 있습니다.
```jsx
const checkIn = function (flightNum, passenger) {
  flightNum = 'LH999'; 
  passenger.name = `Mr. ` + passenger.name;
  if (passenger.passport === 24739479284) {
    alert('Checked in');
  } else {
    alert('Wrong passport!');
  }
};

const flight = 'LH234';
const jonas = {
  name: 'Jonas Schmedtmann',
  passport: 24739479284,
};
checkIn(flight, jonas);

// 얘는 그대론데(real value)
console.log(flight); // LH234
// 얘는 변함(reference)
console.log(jonas); // {name: 'Mr. Jonas Schmedtmann', passport: 2473549124890}
```
